In [14]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression 

In [15]:
df = pd.read_csv('https://sololearn.com/uploads/files/titanic.csv')

In [16]:
df['male'] = df['Sex'] == 'male'

In [17]:
df

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children  \
0           0       3    male  22.0                 1                 0   
1           1       1  female  38.0                 1                 0   
2           1       3  female  26.0                 0                 0   
3           1       1  female  35.0                 1                 0   
4           0       3    male  35.0                 0                 0   
..        ...     ...     ...   ...               ...               ...   
882         0       2    male  27.0                 0                 0   
883         1       1  female  19.0                 0                 0   
884         0       3  female   7.0                 1                 2   
885         1       1    male  26.0                 0                 0   
886         0       3    male  32.0                 0                 0   

        Fare   male  
0     7.2500   True  
1    71.2833  False  
2     7.9250  False  
3    53.1000  False  
4     8.0500   True  
..       ...    ...  
882  13.0000   True  
883  30.0000  False  
884  23.4500  False  
885  30.0000   True  
886   7.7500   True  

[887 rows x 8 columns]

In [18]:
X = df[['Pclass', 'male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].values
# features

In [19]:
X

array([[3, True, 22.0, 1, 0, 7.25],
       [1, False, 38.0, 1, 0, 71.2833],
       [3, False, 26.0, 0, 0, 7.925],
       ...,
       [3, False, 7.0, 1, 2, 23.45],
       [1, True, 26.0, 0, 0, 30.0],
       [3, True, 32.0, 0, 0, 7.75]], dtype=object)

In [24]:
y = df['Survived'].values
#label

In [25]:
y

array([0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,

In [26]:
model = LogisticRegression()
# instantiate the class

In [28]:
model.fit(X, y)
# building the model, using the data to find the line of best fit

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [29]:
print(model.coef_, model.intercept_)

[[-1.1364573  -2.64409955 -0.04237384 -0.38746646 -0.09619818  0.00297184]] [5.08857292]


In [30]:
model.predict(X)

array([0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,

In [32]:
model.predict([[3, True, 22.0, 1, 0, 7.25]])

array([0], dtype=int64)

In [35]:
#check for the first 5 rows
print(model.predict(X[:5]))
# compare it with the target
print(y[:5])

[0 1 1 1 0]
[0 1 1 1 0]


In [38]:
# predicted y values
pred_y = model.predict(X)

In [39]:
# compare it with the target values
y == pred_y

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True, False,
       False,  True,  True, False,  True,  True, False, False,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True, False,  True,  True, False,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True, False,  True,  True,  True,  True,
        True, False,  True,  True,  True,  True,  True,  True, False,
        True, False,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True, False,  True,  True,  True,  True, False,  True,
        True,  True, False,  True, False, False,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True, False,  True,
       False,  True,

In [40]:
# Number of rows for which our model predicts correctly 
print((y == pred_y).sum())

714


This means that of the 887 datapoints, the model makes the correct prediction for 714 of them.

In [42]:
#Total number of passengers 
y.shape[0]

887

In [43]:
#Now we calculate the accuracy of the model and represent it in percentage 
print((y == pred_y).sum() / y.shape[0])

0.8049605411499436


Thus the model’s accuracy is 80%. In other words, the model makes the correct prediction on 80% of the datapoints.


In [45]:
# alternative method for calculating the accuracy
model.score(X,y) 

0.8049605411499436